### PR TITLE
[JIT][Autocast] Don't cast softmax on CPU

### LINKED ISSUE #75866: [JIT] [Autocast] Autocast Pass induce warning in Jit Trace
### 🐛 Describe the bug

We find after we enable the Autocast Pass as default https://github.com/pytorch/pytorch/pull/74178 . There are lots of calculation error warning in `Jit Trace`.

Here is the code to reproduce this error:
```
import torch
import torch.nn as nn

class AtenSoftmaxRepalce(nn.Module):
    def __init__(self, dim=-1):
        super(AtenSoftmaxRepalce, self).__init__()
        self.conv = torch.nn.Conv2d(3, 64, (3, 3), stride=(2, 2), padding=(1, 1), bias=False)
        self.softmax = torch.nn.Softmax(dim)

    def forward(self, x):
        x1 = self.conv(x)
        return self.softmax(x1)

model = AtenSoftmaxRepalce()
model.eval()

x = torch.rand(1, 3, 224, 224).to(torch.bfloat16)

with torch.no_grad():
    with torch.cpu.amp.autocast(cache_enabled=False):
        model = torch.jit.trace(model, x).eval()
```
Here is the details of warning message:
```
/home/lesliefang/pytorch_1_7_1/Autocast_Pytorch/pytorch/torch/jit/_trace.py:983: TracerWarning: Output nr 1. of the traced function does not match the corresponding output of the Python function. Detailed error:
Tensor-likes are not close!

Mismatched elements: 595329 / 802816 (74.2%)
Greatest absolute difference: 0.00012071523815393448 at index (0, 63, 43, 30) (up to 1e-05 allowed)
Greatest relative difference: 0.01244930820610687 at index (0, 27, 106, 34) (up to 1e-05 allowed)

  _check_trace(
```

If we disable the Autocast Pass (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/autocast.cpp#L22), we will not see this warning messages.

### Versions

And we are using stock pytorch with commit: 123297a8c02e0ebbf8b0ae3d3cb16d0dc2e350b4.

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #76661

In eager autocasting, softmax is only up-casted on GPU (and not on CPU).
This fixes the JIT implementation to do the same.

### CODE REVIEW

**Review by jjsjann123 — APPROVED** [review]
LGTM

File: torch/csrc/jit/passes/autocast.cpp [hunk 1]
```diff
@@ -414,6 +413,15 @@ void handleBlock(Block* block, AutocastContext initial_state) {
         }
         break;
 
+      // cast softmax to fp32 only on GPU
+      case aten::softmax:
+        if (!node->schema().is_mutable() && !hasExplicitDtypeArgument(node)) {
+          auto context = current_state();
+          context.cpu_enabled = false;
+          castTensorInputs(node, aten::_autocast_to_full_precision, context);
```
Comment (eellison) [inline torch/csrc/jit/passes/autocast.cpp:421]: isnt the modification of `cpu_enabled` going to persist pass this op?

Comment (davidberard98) [reply torch/csrc/jit/passes/autocast.cpp:421]: I think since I'm copying the state to `auto context` (i.e. not by reference) it's not going to affect other ops

**Review by eellison — APPROVED** [review]
